nixos/tachidesk-server: init at 0.7.0

RatCornu opened the pull request:
## Description of changes

https://github.com/Suwayomi/Tachidesk-Server

A free and open source manga reader server that runs extensions built for [Tachiyomi](https://tachiyomi.org/).

Tested the program with different configurations on an headless server. 

Added a nixos module and a package called in the module.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

RatCornu committed:
tachidesk: init at 0.7.0

RatCornu committed:
maintainers: add ratcornu

RatCornu committed:
nixos/tachidesk-server: init at 0.7.0

h7x4 commented:
I think you'll have to reorder your commits, so you add yourself to the maintainer list before initializing the package

RatCornu commented:
Indeed, I've just reordered them, thanks !

h7x4 reviewed (changes requested):
Cool stuff! Just a bit of nits here and there.

I think it would be nice to have a small nixos test to verify that the module runs with a minimal config, but I'm not planning to block the PR if you don't wanna deal with that.

RatCornu committed:
nixos/tachidesk-server: add nixos tests

RatCornu commented on pkgs/by-name/ta/tachidesk-server/package.nix:
```diff
@@ -0,0 +1,41 @@
+
+  src = fetchurl {
+    url = "https://github.com/Suwayomi/Tachidesk-Server/releases/download/v0.7.0/Tachidesk-Server-v${finalAttrs.version}-r${toString finalAttrs.revision}.jar";
+    sha256 = "sha256-4DO1WiBCu/8ypFgJdBmEwQXQ1xaWAlbt8N5TELomVVA=";
```
Done!

RatCornu commented on nixos/modules/services/web-apps/tachidesk-server.nix:
```diff
@@ -0,0 +1,200 @@
+        Type = "simple";
+        Restart = "on-failure";
+        ExecStartPre = [
+          "${pkgs.coreutils}/bin/mkdir -p ${cfg.dataDir}/.local/share/Tachidesk"
```
I am not sure on how to use it, I tried something and it works so I guess it's okay

RatCornu commented on nixos/modules/services/web-apps/tachidesk-server.nix:
```diff
@@ -0,0 +1,200 @@
+            else if attrType == "bool" then boolToString attr
+            else ''${toString attr}''
+        );
+        configFile = pkgs.writeText "server.conf" (strings.concatStrings [
```
I added a comment and change the function name to `toHOCON`

RatCornu commented on nixos/modules/services/web-apps/tachidesk-server.nix:
```diff
@@ -0,0 +1,200 @@
+        };
+
+        localSourcePath = mkOption {
+          type = types.str;
```
Done!

RatCornu commented on nixos/modules/services/web-apps/tachidesk-server.nix:
```diff
@@ -0,0 +1,200 @@
+            '';
+          };
+
+          password = mkOption {
```
I tried something but I am not sure about my solution, does it seem okay to you ?

RatCornu commented on nixos/modules/services/web-apps/tachidesk-server.nix:
```diff
@@ -0,0 +1,200 @@
+
+          username = mkOption {
+            type = types.str;
+            default = "username";
```
Is it okay like this or did I get something wrong ?

RatCornu commented on nixos/modules/services/web-apps/tachidesk-server.nix:
```diff
@@ -0,0 +1,200 @@
+      };
+
+      dataDir = mkOption {
+        type = types.str;
```
Done!

RatCornu commented on nixos/modules/services/web-apps/tachidesk-server.nix:
```diff
@@ -0,0 +1,200 @@
+        type = types.package;
+        description = mdDoc "Which package to use for the Tachidesk-Server instance";
+        relatedPackages = [ "tachidesk-server" ];
+      };
```
Done!

RatCornu commented on nixos/modules/services/web-apps/tachidesk-server.nix:
```diff
@@ -0,0 +1,200 @@
+        default = {};
+        description = mdDoc ''
+          Extra configuration to write in the file `server.conf`.
+          See [https://github.com/Suwayomi/Tachidesk-Server/wiki/Configuring-Tachidesk-Server] for more informations.
```
Done!

RatCornu commented on nixos/modules/services/web-apps/tachidesk-server.nix:
```diff
@@ -0,0 +1,200 @@
+{ config, pkgs, lib, ... }:
+
+with lib;
```
Done!

RatCornu commented on nixos/modules/services/web-apps/tachidesk-server.md:
```diff
@@ -0,0 +1,84 @@
+# Tachidesk-server {#module-services-tachidesk-server}
```
I like documentation

RatCornu commented on nixos/modules/services/web-apps/tachidesk-server.nix:
```diff
@@ -0,0 +1,200 @@
+
+        localSourcePath = mkOption {
+          type = types.str;
+          default = "";
```
Indeed, `dataDir` can be use as default `localSourcePath`

RatCornu commented on pkgs/by-name/ta/tachidesk-server/package.nix:
```diff
@@ -0,0 +1,41 @@
+  ];
+
+  buildCommand = ''
+    makeWrapper ${jdk17_headless}/bin/java $out/bin/tachidesk-server --add-flags "-jar $src"
```
Done!

RatCornu commented on pkgs/by-name/ta/tachidesk-server/package.nix:
```diff
@@ -0,0 +1,41 @@
+    homepage = "https://github.com/Suwayomi/Tachidesk-Server";
+    license = licenses.mpl20;
+    platforms = jdk17_headless.meta.platforms;
+    maintainers = with maintainers; [ ratcornu ];
```
Done!

RatCornu commented on nixos/modules/services/web-apps/tachidesk-server.nix:
```diff
@@ -0,0 +1,200 @@
+        configFile = pkgs.writeText "server.conf" (strings.concatStrings [
+          ''
+            server.ip = "${cfg.config.ip}"
+            server.port = ${toString cfg.config.port}
```
Done!

RatCornu commented on nixos/modules/services/web-apps/tachidesk-server.nix:
```diff
@@ -0,0 +1,200 @@
+          in if attrType == "string" then ''"${attr}"''
+            else if attrType == "int" then ''${toString attr}''
+            else if attrType == "bool" then boolToString attr
+            else ''${toString attr}''
```
Done!

RatCornu commented on nixos/modules/services/web-apps/tachidesk-server.nix:
```diff
@@ -0,0 +1,200 @@
+            concatLists (mapAttrsToList (name: value:
+              optional (value != null) ''${name} = ${mapAttrToString value}''
+            ) cfg.extraConfig)
+          )))
```
Done!

RatCornu commented:
> Cool stuff! Just a bit of nits here and there.
> 
> I think it would be nice to have a small nixos test to verify that the module runs with a minimal config, but I'm not planning to block the PR if you don't wanna deal with that.

I added a minimal nixos test to check if the port is open !

h7x4 commented:
Cool. Feel free to mark the relevant comments as resolved. I'll have another look through it soon!

RatCornu commented on nixos/modules/services/web-apps/tachidesk-server.nix:
```diff
@@ -0,0 +1,200 @@
+        '';
+      };
+
+      user = mkOption {
```
I think keeping `dataDir` can be useful as it will be used to store manga scans downloaded

h7x4 commented:
I think `nixos/tachidesk-server: change password option to passwordFile` should probably be merged into `nixos/tachidesk-server: init at 0.7.0`

h7x4 commented on nixos/modules/services/web-apps/tachidesk-server.nix:
```diff
@@ -0,0 +1,211 @@
+        Restart = "on-failure";
+        EnvironmentFile = "";
+        ExecStartPre = [
+          "${pkgs.envsubst}/bin/envsubst -i ${configFile} -o ${tmpConfigFile}"
+          "${pkgs.coreutils}/bin/cp ${tmpConfigFile} ${cfg.dataDir}/.local/share/Tachidesk/server.conf"
+          "${pkgs.coreutils}/bin/rm ${tmpConfigFile}"
```
Maybe this could just substitute straight into `dataDir`, without needing to go via `/tmp`?

RatCornu replied:
I tried but it seemed that envsubst wouldn't work directly by putting `dataDir` directly. I double-check this

h7x4 replied:
Well, if it ends up not working and this seems like the best workaround, it would probably warrant a comment. Without any explanation, it seems really unintuitive.

RatCornu replied:
Okay it seems to work, so I guess I did something wrong ^^'

h7x4 commented:
I'm having some trouble running the test.

```console
[nix-shell:~/.cache/nixpkgs-review/pr-257733/nixpkgs]$ nix-build -A nixosTests.tachidesk-server
error: attribute 'dataDir' missing

       at /home/h7x4/.cache/nixpkgs-review/pr-257733/nixpkgs/nixos/modules/services/web-apps/tachidesk-server.nix:97:21:

           96|           type = types.path;
           97|           default = cfg.config.dataDir;
             |                     ^
           98|           defaultText = lib.literalExpression "config.dataDir";
(use '--show-trace' to show detailed location information)

```

RatCornu commented:
Yeah I'm sorry, I force pushed several times so I think you do not have the latest version of my branch : i added a `defaultText` to fix this

RatCornu commented:
I am sorry to ask but since 23.11 is near its release, do I have to change something on this PR ?

mrcjkb commented on nixos/modules/services/web-apps/tachidesk-server.nix:
```diff
@@ -0,0 +1,208 @@
+          };
+
+          passwordFile = mkOption {
+            type = types.path;
```
question: should this have a default option (if there's no appropriate default, can the type be a union with null)?

ditto for `username`?

RatCornu replied:
I don't think putting a default is a good option here so I change the type to `nullOr path`.

mrcjkb replied:
I suppose you could make the default `null` and throw an error if it is `null`?

RatCornu replied:
Oh yeah it makes sense, I change this!

mrcjkb commented on nixos/modules/services/web-apps/tachidesk-server.nix:
```diff
@@ -0,0 +1,208 @@
+        };
+
+        basicAuth = {
+          enable = mkOption {
```
suggestion: use `lib.mkEnableOption`

RatCornu replied:
Indeed it is better suited

mrcjkb commented on nixos/modules/services/web-apps/tachidesk-server.nix:
```diff
@@ -0,0 +1,208 @@
+    };
+  };
+
+  meta.maintainers = with lib.maintainers; [ ratcornu ];
```
suggestion: add `meta.doc`

RatCornu replied:
Indeed, it's added

mrcjkb reviewed (changes requested):
#### Module

##### Reviewed points

- [x] module path fits the guidelines
- [x] options have appropriate types
- [ ] options have default
- [ ] options have example
- [x] options have descriptions
- [x] No unneeded package is added to `environment.systemPackages`
- [x] `meta.maintainers` is set
- [ ] module documentation is declared in `meta.doc`

##### Possible improvements

- Add defaults (if applicable) and examples to all module options


##### Comments

As this adds a new module, it should probably be included in the release notes.

#### Package

##### Reviewed points

- [x] package path fits guidelines
- [x] package name fits guidelines
- [x] package version fits guidelines
- [x] `meta.description` is set and fits guidelines
- [x] `meta.license` fits upstream license
- [x] `meta.platforms` is set
- [x] `meta.maintainers` is set
- [x] build time only dependencies are declared in `nativeBuildInputs`
- [x] source is fetched using the appropriate function
- [x] the list of `phases` is not overridden

🚀🚀🚀 for the documentation and NixOS test

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/3175

mrcjkb commented:
Result of `nixpkgs-review pr 257733` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>tachidesk-server</li>
  </ul>
</details>

mrcjkb commented:
#### Module

##### Reviewed points

- [x] module path fits the guidelines
- [x] module tests succeed on ARCHITECTURE
- [x] options have appropriate types
- [ ] options have default -> not all
- [ ] options have example
- [x] options have descriptions
- [x] No unneeded package is added to `environment.systemPackages`
- [x] `meta.maintainers` is set
- [x] module documentation is declared in `meta.doc`

##### Possible improvements

- Add missing defaults to options (null if no default makes sense)
- Add examples to options

##### Comments

- Since it's a new module, there should probably be
  an entry in the release notes.

#### Package

##### Reviewed points

- [x] package name fits guidelines
- [x] package version fits guidelines
- [x] package builds on x86_64-linux
- [x] executables tested on x86_64-linux
- [x] all depending packages build

##### Comments

I ran the executable and got this error:

```console
java.lang.UnsatisfiedLinkError: Can't load library: /nix/store/iym0anb6r2r3iqflwvmsr75a4yk40sv6-openjdk-headless-17.0.7+7/lib/openjdk/lib/libawt_xawt.so
        at java.base/java.lang.ClassLoader.loadLibrary(ClassLoader.java:2393)
        at java.base/java.lang.Runtime.load0(Runtime.java:755)

# rest of the stack trace omitted for brevity
```

RatCornu commented:
> Somehow, my previous comment appears to have been deleted. So here's a new one based on it (with some updates)
> #### Module
> ##### Reviewed points
> 
>     * [x]  module path fits the guidelines
> 
>     * [x]  options have appropriate types
> 
>     * [ ]  options have default _-> not all of them_
> 
>     * [ ]  options have example
> 
>     * [x]  options have descriptions
> 
>     * [x]  No unneeded package is added to `environment.systemPackages`
> 
>     * [x]  `meta.maintainers` is set
> 
>     * [x]  module documentation is declared in `meta.doc`
> 
> 
> ##### Possible improvements
> 
>     * Add missing defaults to options (null if no default makes sense)
> 
>     * Add examples to options
> 
> 
> ##### Comments
> 
>     * Since it's a new module, there should probably be
>       an entry in the release notes.
> 
> 
> #### Package
> ##### Reviewed points
> 
>     * [x]  package name fits guidelines
> 
>     * [x]  package version fits guidelines
> 
>     * [x]  package builds on x86_64-linux
> 
>     * [x]  executables tested on x86_64-linux
> 
>     * [x]  all depending packages build
> 
> 
> ##### Comments
> 
> I ran the executable and got this error:
> 
> ```
> java.lang.UnsatisfiedLinkError: Can't load library: /nix/store/iym0anb6r2r3iqflwvmsr75a4yk40sv6-openjdk-headless-17.0.7+7/lib/openjdk/lib/libawt_xawt.so
>         at java.base/java.lang.ClassLoader.loadLibrary(ClassLoader.java:2393)
>         at java.base/java.lang.Runtime.load0(Runtime.java:755)
> 
> # rest of the stack trace omitted for brevity
> ```

Oh I don't have this issue on my server (with the module enabled), I try to figure out where this come from.

RatCornu commented:
Could you send me the commands you ran to obtain this result ? With the following commands:

```
$ nix-build -A tachidesk-server

$ ./result/bin/tachidesk-server
```
I don't get any error

mrcjkb commented:
> /result/bin/tachidesk-server

Yup, that's exactly what I ran.

<details>
<summary>
Here's the full log
</summary>

```console
> nix-build -A tachidesk-server
> ./result/bin/tachidesk-server

23:46:41.600 [main] INFO  suwayomi.tachidesk.server.ServerSetup - Running Tachidesk v0.7.0 revision r1197
23:46:41.799 [main] INFO  suwayomi.tachidesk.server.util.AppMutex - Mutex status is clear, Resuming startup.
23:46:42.018 [main] INFO  suwayomi.tachidesk.server.database.DBManager - Using h2 database version 1.4
23:46:42.153 [main] INFO  de.neonew.exposed.migrations.RunMigrations - Running migrations on database jdbc:h2:/home/mrcjk/.local/share/Tachidesk/database
23:46:42.423 [main] INFO  de.neonew.exposed.migrations.RunMigrations - Database version before migrations: 25
23:46:42.424 [main] INFO  de.neonew.exposed.migrations.RunMigrations - Migrations finished successfully
java.lang.ExceptionInInitializerError
        at dorkbox.jna.linux.GtkEventDispatch.startGui(GtkEventDispatch.java:80)
        at dorkbox.systemTray.SystemTray.get(SystemTray.java:651)
        at suwayomi.tachidesk.server.util.SystemTray.systemTray(SystemTray.kt:31)
        at suwayomi.tachidesk.server.ServerSetupKt$systemTrayInstance$2.invoke(ServerSetup.kt:54)
        at suwayomi.tachidesk.server.ServerSetupKt$systemTrayInstance$2.invoke(ServerSetup.kt:54)
        at kotlin.SynchronizedLazyImpl.getValue(LazyJVM.kt:74)
        at suwayomi.tachidesk.server.ServerSetupKt.getSystemTrayInstance(ServerSetup.kt:54)
        at suwayomi.tachidesk.server.ServerSetupKt.applicationSetup(ServerSetup.kt:145)
        at suwayomi.tachidesk.MainKt.main(Main.kt:14)
        at suwayomi.tachidesk.MainKt.main(Main.kt)
Caused by: java.lang.RuntimeException: We apologize for this, but we are unable to determine the GTK library is in use, or even if it is in use... Please create an issue for this and include your OS type and configuration.
        at dorkbox.jna.linux.GtkLoader.<clinit>(GtkLoader.java:260)
        ... 10 more
23:46:42.696 [main] INFO  suwayomi.tachidesk.server.util.WebInterfaceManager - WebUI Static files exists and is the correct revision
23:46:42.697 [main] INFO  suwayomi.tachidesk.server.util.WebInterfaceManager - Verifying WebUI Static files...
23:46:42.721 [main] INFO  suwayomi.tachidesk.server.util.WebInterfaceManager - md5: 1af53126be02cba56e539444603b1a09
23:46:42.721 [main] INFO  suwayomi.tachidesk.server.JavalinSetup - Serving web static files for WebUI
23:46:42.744 [main] WARN  io.javalin.Javalin - JDK15 and Kotlin 1.5 break reflection in different ways - the OpenAPI plugin will not work properly. Please visit https://github.com/tipsy/javalin/issues/1193 if you want to help fix this issue.
23:46:42.744 [main] WARN  io.javalin.Javalin - You can disable this warning by doing `OpenApiVersionUtil.logWarnings = false`
23:46:42.753 [main] INFO  io.javalin.Javalin -
       __                      __ _            __ __
      / /____ _ _   __ ____ _ / /(_)____      / // /
 __  / // __ `/| | / // __ `// // // __ \    / // /_
/ /_/ // /_/ / | |/ // /_/ // // // / / /   /__  __/
\____/ \__,_/  |___/ \__,_//_//_//_/ /_/      /_/

          https://javalin.io/documentation

23:46:42.753 [main] INFO  io.javalin.Javalin - Starting Javalin ...
23:46:42.760 [main] INFO  io.javalin.Javalin - You are running Javalin 4.6.6 (released October 2, 2022. Your Javalin version is 454 days old. Consider checking for a newer version.).
23:46:42.823 [main] INFO  io.javalin.Javalin - Static file handler added: StaticFileConfig(hostedPath=/, directory=/home/mrcjk/.local/share/Tachidesk/webUI, location=EXTERNAL, precompress=false, aliasCheck=null, headers={Cache-Control=max-age=0}). File system location: '/home/mrcjk/.local/share/Tachidesk/webUI'
23:46:42.823 [main] INFO  io.javalin.Javalin - Static file handler added: StaticFileConfig(hostedPath=/, directory=META-INF/resources/webjars, location=CLASSPATH, precompress=false, aliasCheck=null, headers={Cache-Control=max-age=31622400}). File system location: 'jar:file:/nix/store/n1y01rrhx2s3jdnk8x2c806rsw1jsprl-Tachidesk-Server-v0.7.0-r1197.jar!/META-INF/resources/webjars'
23:46:42.823 [main] INFO  io.javalin.Javalin - Listening on http://0.0.0.0:4567/
23:46:42.823 [main] INFO  io.javalin.Javalin - Javalin started in 70ms \o/
java.lang.UnsatisfiedLinkError: Can't load library: /nix/store/y33fifmr59gzsswsvradpkskzz62xczx-openjdk-headless-17.0.7+7/lib/openjdk/lib/libawt_xawt.so
        at java.base/java.lang.ClassLoader.loadLibrary(ClassLoader.java:2393)
        at java.base/java.lang.Runtime.load0(Runtime.java:755)
        at java.base/java.lang.System.load(System.java:1953)
        at java.base/jdk.internal.loader.NativeLibraries.load(Native Method)
        at java.base/jdk.internal.loader.NativeLibraries$NativeLibraryImpl.open(NativeLibraries.java:388)
        at java.base/jdk.internal.loader.NativeLibraries.loadLibrary(NativeLibraries.java:232)
        at java.base/jdk.internal.loader.NativeLibraries.loadLibrary(NativeLibraries.java:174)
        at java.base/jdk.internal.loader.NativeLibraries.findFromPaths(NativeLibraries.java:315)
        at java.base/jdk.internal.loader.NativeLibraries.loadLibrary(NativeLibraries.java:285)
        at java.base/java.lang.ClassLoader.loadLibrary(ClassLoader.java:2398)
        at java.base/java.lang.Runtime.loadLibrary0(Runtime.java:818)
        at java.base/java.lang.System.loadLibrary(System.java:1989)
        at java.desktop/java.awt.Toolkit$2.run(Toolkit.java:1388)
        at java.desktop/java.awt.Toolkit$2.run(Toolkit.java:1386)
        at java.base/java.security.AccessController.doPrivileged(AccessController.java:318)
        at java.desktop/java.awt.Toolkit.loadLibraries(Toolkit.java:1385)
        at java.desktop/java.awt.Toolkit.initStatic(Toolkit.java:1423)
        at java.desktop/java.awt.Toolkit.<clinit>(Toolkit.java:1397)
        at java.desktop/java.awt.Desktop.isDesktopSupported(Desktop.java:329)
        at dorkbox.desktop.Desktop.awtSupported(Desktop.java:297)
        at dorkbox.desktop.Desktop.browseURL(Desktop.java:146)
        at dorkbox.desktop.Desktop.browseURL(Desktop.java:122)
        at suwayomi.tachidesk.server.util.Browser.openInBrowser(Browser.kt:30)
        at suwayomi.tachidesk.server.JavalinSetup.javalinSetup$lambda$3$lambda$2(JavalinSetup.kt:75)
        at io.javalin.core.event.EventManager.fireEvent(EventManager.kt:18)
        at io.javalin.Javalin.start(Javalin.java:176)
        at io.javalin.Javalin.start(Javalin.java:148)
        at io.javalin.Javalin.start(Javalin.java:134)
        at suwayomi.tachidesk.server.JavalinSetup.javalinSetup(JavalinSetup.kt:78)
        at suwayomi.tachidesk.MainKt.main(Main.kt:15)
        at suwayomi.tachidesk.MainKt.main(Main.kt)
^[^C23:46:48.317 [Thread-14] INFO  io.javalin.Javalin - Stopping Javalin ...
23:46:48.322 [Thread-14] INFO  io.javalin.Javalin - Javalin has stopped

```
</details>

mrcjkb commented:
I just tried replacing `jdk17_headless` with `jdk17`, and it runs fine - but it opens a browser window.

mrcjkb commented:
:thinking: I was able to fix it by adding a `-Dsuwayomi.tachidesk.config.server.initialOpenInBrowserEnabled=false` flag to the wrapper:

```nix
  buildCommand = ''
    makeWrapper ${jdk}/bin/java $out/bin/tachidesk-server --add-flags "-Dsuwayomi.tachidesk.config.server.initialOpenInBrowserEnabled=false -jar $src"
  '';

```

RatCornu commented:
Thanks! I also found a fix by adding `-Djava.awt.headless=true` to the JVM args. What do you think better suited ?

mrcjkb commented:
> Thanks! I also found a fix by adding `-Djava.awt.headless=true` to the JVM args. What do you think better suited ?

I tried that, and it got rid of the linker error, but it still tried to open a browser window.

RatCornu commented:
Okay! I'll work on those changes as soon as possible

RatCornu commented:
I fixed the small issues I found after testing, and I didn't manage to make the package run with the non-headless `jdk17`. I got a SIGSEGV because of GTK. As the program is intended to run in the browser and can run without issue with the `jdk17_headless`, I think it is not a big issue if `initialOpenInBrowserEnabled` is always set to `false`.

Moreover, I added an option `enableSystemTray` which is recommended by the wiki for headless servers.

SuperSandro2000 commented on nixos/modules/services/web-apps/tachidesk-server.nix:
```diff
@@ -0,0 +1,227 @@
+
+          (
+            if cfg.config.basicAuth.enable then
+              (if builtins.isNull cfg.config.basicAuth.username || builtins.isNull cfg.config.basicAuth.passwordFile then
```
```suggestion
              (if (cfg.config.basicAuth.username == null || cfg.config.basicAuth.passwordFile == null) then
```
explizit comparisons are easier to read

RatCornu replied:
I was not sure if this worked so I used the function instead, but it's now changed

SuperSandro2000 commented on nixos/modules/services/web-apps/tachidesk-server.nix:
```diff
@@ -0,0 +1,227 @@
+        after = [ "network-online.target" ];
+
+        environment = {
+          TACHIDESK_SERVER_BASIC_AUTH_PASSWORD = if cfg.config.basicAuth.enable then "${lib.fileContents cfg.config.basicAuth.passwordFile}" else "";
```
```suggestion
          TACHIDESK_SERVER_BASIC_AUTH_PASSWORD = lib.optionalString (!cfg.config.basicAuth.enable) "${lib.fileContents cfg.config.basicAuth.passwordFile}";
```
we have a shorthand for that

RatCornu replied:
It's changed

SuperSandro2000 commented on nixos/modules/services/web-apps/tachidesk-server.nix:
```diff
@@ -0,0 +1,227 @@
+
+          Type = "simple";
+          Restart = "on-failure";
+          EnvironmentFile = "";
```
```suggestion
```
we don't need to set that empty

RatCornu replied:
Indeed, it's removed

SuperSandro2000 commented on nixos/modules/services/web-apps/tachidesk-server.nix:
```diff
@@ -0,0 +1,227 @@
+          Restart = "on-failure";
+          EnvironmentFile = "";
+          ExecStartPre = [
+            "${pkgs.envsubst}/bin/envsubst -i ${configFile} -o ${cfg.dataDir}/.local/share/Tachidesk/server.conf"
+          ];
+          ExecStart = "${cfg.package}/bin/tachidesk-server -Dsuwayomi.tachidesk.config.server.rootDir=${cfg.dataDir}";
```
```suggestion
            "${lib.getExe pkgs.envsubst} -i ${configFile} -o ${cfg.dataDir}/.local/share/Tachidesk/server.conf"
          ];
          ExecStart = "${lib.getExe cfg.package} -Dsuwayomi.tachidesk.config.server.rootDir=${cfg.dataDir}";
```

RatCornu replied:
I was not aware that this existed, it is changed

SuperSandro2000 commented on pkgs/by-name/ta/tachidesk-server/package.nix:
```diff
@@ -0,0 +1,52 @@
+  revision = 1197;
+
+  src = fetchurl {
+    url = "https://github.com/Suwayomi/Tachidesk-Server/releases/download/v0.7.0/Tachidesk-Server-v${finalAttrs.version}-r${toString finalAttrs.revision}.jar";
```
```suggestion
    url = "https://github.com/Suwayomi/Tachidesk-Server/releases/download/v${finalAttrs.version}/Tachidesk-Server-v${finalAttrs.version}-r${toString finalAttrs.revision}.jar";
```

RatCornu replied:
Done

SuperSandro2000 commented on pkgs/by-name/ta/tachidesk-server/package.nix:
```diff
@@ -0,0 +1,52 @@
+  ];
+
+  buildCommand = ''
+    makeWrapper ${jdk}/bin/java $out/bin/tachidesk-server --add-flags "-Dsuwayomi.tachidesk.config.server.initialOpenInBrowserEnabled=false -jar $src"
```
```suggestion
    makeWrapper ${jdk}/bin/java $out/bin/tachidesk-server \
      --add-flags "-Dsuwayomi.tachidesk.config.server.initialOpenInBrowserEnabled=false -jar $src"
```
we don't have a strict line length but doing a line break here is common to avoid long lines

RatCornu replied:
I'll remember it for my future PR, it's changed

SuperSandro2000 commented on pkgs/by-name/ta/tachidesk-server/package.nix:
```diff
@@ -0,0 +1,52 @@
+  meta = with lib; {
+    description = "A free and open source manga reader server that runs extensions built for Tachiyomi.";
+    longDescription = ''
+      A free and open source manga reader server that runs extensions built for Tachiyomi.
+
```
```suggestion
```
we don't need to repeat the description again, as it is usually right next to each other anyway.

RatCornu replied:
Done

RatCornu commented on nixos/modules/services/web-apps/tachidesk-server.md:
```diff
@@ -0,0 +1,105 @@
+
+## Basic usage {#module-services-tachidesk-server-basic-usage}
+
+By default, the module will execute tachidesk-server backend and webUI by enabling the `tachidesk-server` service:
```
It's changed.

RatCornu commented on nixos/modules/services/web-apps/tachidesk-server.md:
````diff
@@ -0,0 +1,105 @@
+
+```
+{ ... }:
+
````
Done for every code block

RatCornu commented on nixos/modules/services/web-apps/tachidesk-server.md:
````diff
@@ -0,0 +1,105 @@
+}
+```
+
+Beware, the password will be written in clear in your nix store.
````
Indeed, and I changed the `password` option that does not exist anymore for `passwordFile`.

RatCornu commented on nixos/modules/services/web-apps/tachidesk-server.nix:
```diff
@@ -0,0 +1,227 @@
+
+        basicAuth = {
+          enable = mkEnableOption (mdDoc ''
+            Controls whether if Tachidesk-Server should be protected with Basic access authentication.
```
I'm not sure if I did this right, but I changed the description

RatCornu commented on nixos/modules/services/web-apps/tachidesk-server.nix:
```diff
@@ -0,0 +1,227 @@
+        };
+      };
+
+      extraConfig = mkOption {
```
I never used a freeform type so I'm not sure but I changed it and I think I did it right

RatCornu commented:
I think I resolved everything that was asked and I tested all the changes on my server

mrcjkb commented:
> I think I resolved everything that was asked and I tested all the changes on my server

On my end, these two points are still open:

- [ ] Add a the new module to the [24.05 release notes](https://github.com/NixOS/nixpkgs/blob/3cb442f49442e3dc76ecc661fbe236362396da8e/nixos/doc/manual/release-notes/rl-2405.section.md).
- [ ] Add `example` attributes to the module options.

RatCornu commented:
> > I think I resolved everything that was asked and I tested all the changes on my server
> 
> On my end, these two points are still open:
> 
>     * [ ]  Add a the new module to the [24.05 release notes](https://github.com/NixOS/nixpkgs/blob/3cb442f49442e3dc76ecc661fbe236362396da8e/nixos/doc/manual/release-notes/rl-2405.section.md).
> 
>     * [ ]  Add `example` attributes to the module options.

Oh I'm sorry I was talking about the points of the other review. I forgot to do this, I'll do it now: should I add `example` to all module options? And I will search how to include this module in the release notes

mrcjkb commented:
> should I add example to all module options?

Tbh, I don't know if it's a hard requirement. I don't think it's necessary for trivial (e.g. boolean) options, but otherwise I would say it's a good idea.

RatCornu committed:
nixos/tachidesk-server: add release note

RatCornu commented:
Okay I added examples where it seems needed, and I added the release note.

RatCornu commented:
@SuperSandro2000 sorry to ping you: is it okay for you as well or should I make final changes ? \^\^

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/1379

RatCornu commented:
> Cool stuff! Just a bit of nits here and there.
> 
> I think it would be nice to have a small nixos test to verify that the module runs with a minimal config, but I'm not planning to block the PR if you don't wanna deal with that.

A minimal test have been added in the module so that it checks that the given port is listening after the start.

h7x4 commented on nixos/modules/services/web-apps/tachidesk-server.nix:
```diff
@@ -0,0 +1,235 @@
+        after = [ "network-online.target" ];
+
+        environment = {
+          TACHIDESK_SERVER_BASIC_AUTH_PASSWORD = lib.optionalString (cfg.config.basicAuth.enable) "${lib.fileContents cfg.config.basicAuth.passwordFile}";
```
This will result in IFD which is not allowed in nixpkgs, and it will also lead to the password being included in the nix store for anyone to see.

Alternative ways of doing it includes:

- Reading the contents of the file into the envvar during systemd's `Exec` phase (`serviceConfig.script` might be useful, [here's an example](https://github.com/NixOS/nixpkgs/blob/cec578e2b429bf59855063760d668cae355adb6d/nixos/modules/services/web-apps/nextcloud-notify_push.nix#L90)).
- Defining an `EnvironmentFile` option, and add it to the systemd unit. Add the name of the important envvars to the description of the option.
- Asking upstream to support reading files directly

RatCornu replied:
Indeed I didn't pay attention. I change to make the script looks like your example.

h7x4 commented on nixos/modules/services/web-apps/tachidesk-server.nix:
```diff
@@ -0,0 +1,235 @@
+          ''
+
+          (
+            if cfg.config.basicAuth.enable then
+              (if cfg.config.basicAuth.username == null || cfg.config.basicAuth.passwordFile == null then
+                throw ''
+                  [tachidesk-server]: the username and the password file cannot be null when the basic auth is enable
```
I think these kinds of warnings are better suited for `assertions`. They aren't directly related to the configuration as much as it is a prerequisite for being able to configure the service.

```nix
assertions = [{
  assertion = cfg.config.basicAuth.enable -> (cfg.config.basicAuth.username != null && cfg.config.basicAuth.passwordFile != null);
  message = ''
    [tachidesk-server]: the username and the password file cannot be null when the basic auth is enable
  '';
}];
```

RatCornu replied:
I didn't know this syntax but I like it: I will surely reuse this. It's done

h7x4 commented on nixos/modules/services/web-apps/tachidesk-server.nix:
```diff
@@ -0,0 +1,235 @@
+{
+  options = {
+    services.tachidesk-server = {
+      enable = mkEnableOption (mdDoc "A free and open source manga reader server that runs extensions built for Tachiyomi.");
```
`mkEnableOption` automatically prepends "Whether to enable", a capitalized `A` would seem a bit out of place.

```suggestion
      enable = mkEnableOption (mdDoc "Tachidesk, a free and open source manga reader server that runs extensions built for Tachiyomi.");
```

RatCornu replied:
Done

h7x4 commented on nixos/modules/services/web-apps/tachidesk-server.nix:
```diff
@@ -0,0 +1,235 @@
+
+      extraConfig = mkOption {
+        type = types.submodule {
+          freeformType = with lib.types; attrsOf str;
```
The example you provide below doesn't match up with the type. According to the values the HOCON generator seems to accept, I think the `freeformType` should be more accepting.

```suggestion
          freeformType = with lib.types; attrsOf (oneOf [
            str
            path
            int
            float
            bool
          ])
```

RatCornu replied:
Indeed it was the first time I used free form, I wasn't sure of the syntax. It's done

h7x4 commented on nixos/modules/services/web-apps/tachidesk-server.nix:
```diff
@@ -0,0 +1,235 @@
+    users.users = mkIf (cfg.user == "tachidesk") {
+      tachidesk = {
+        group = cfg.group;
+        home = cfg.dataDir;
```
I do not think a system user should have a home directory. As long as the datadir is readable and writable by the user, that's enough.

```suggestion
```

RatCornu replied:
Done

h7x4 commented on nixos/modules/services/web-apps/tachidesk-server.nix:
```diff
@@ -0,0 +1,235 @@
+        type = types.bool;
+        default = false;
+        description = mdDoc ''
+          A System Tray icon will be tried to be created.
```
nit:

```suggestion
          Whether to enable a system tray icon, if possible.
```

RatCornu replied:
Done

h7x4 reviewed (commented):
I have a final few comments, hope you don't mind :)

RatCornu commented:
> I have a final few comments, hope you don't mind :)

No problem \^\^ I'm happy this MR is not forgotten x).
I will solve this tomorrow!

h7x4 commented on pkgs/by-name/ta/tachidesk-server/package.nix:
```diff
@@ -0,0 +1,51 @@
+    makeWrapper
+  ];
+
+  buildCommand = ''
+    makeWrapper ${jdk}/bin/java $out/bin/tachidesk-server \
+      --add-flags "-Dsuwayomi.tachidesk.config.server.initialOpenInBrowserEnabled=false -jar $src"
+  '';
```
Just making sure, was there a good reason for the choice of `buildCommand` over `buildPhase`? This seems to disable all other phases, making overrides with extensions to `patchPhase` or `buildPhase` completely useless without warning.

Quoting the [nixpkgs manual about stdenv phases](https://nixos.org/manual/nixpkgs/unstable/#sec-stdenv-phases):

> This generic command (mkDerivation) either invokes a script at buildCommandPath, or a buildCommand, or a number of phases. Package builds are split into phases to make it easier to override specific parts of the build (e.g., unpacking the sources or installing the binaries).

RatCornu replied:
> Just making sure, was there a good reason for the choice of `buildCommand` over `buildPhase`? This seems to disable all other phases, making overrides with extensions to `patchPhase` or `buildPhase` completely useless without warning.
> 
> Quoting the [nixpkgs manual about stdenv phases](https://nixos.org/manual/nixpkgs/unstable/#sec-stdenv-phases):
> 
> > This generic command (mkDerivation) either invokes a script at buildCommandPath, or a buildCommand, or a number of phases. Package builds are split into phases to make it easier to override specific parts of the build (e.g., unpacking the sources or installing the binaries).

As I didn't know very well what to do, I copied the [komga package](https://github.com/NixOS/nixpkgs/blob/301f4bfbd7c7d5a7cec710f91506c6eda4d32f94/pkgs/servers/komga/default.nix#L22) which seemed really close to me. I'm not sure this is the best option to choose

h7x4 replied:
Then I think it would be better if you wrote it like this (same goes for komga, but let's take that another time):

```
# I think you might have to do this, so it doesn't try to unzip the jar
dontUnpack = true;

buildPhase = ''
  runHook preBuild
  
  makeWrapper ${jdk}/bin/java $out/bin/tachidesk-server \
    --add-flags "-Dsuwayomi.tachidesk.config.server.initialOpenInBrowserEnabled=false -jar $src"
  
  runHook postBuild
'';
```

h7x4 commented:
Also, would you mind moving most of the review stuff from the release notes commit to the nixos/tachidesk-server commit?

RatCornu commented:
> Also, would you mind moving most of the review stuff from the release notes commit to the nixos/tachidesk-server commit?

Indeed it is more coherent ^^', I just fix a liitle problem on my tests and I do this.

RatCornu commented:
Okay I did three things :

- I chaged `buildCommand` to `buildPhase` as you suggested
- I put the changes in the correct commits
- I put `users.users.tachidesk.home = cfg.dataDir` back because the program `tachidesk-server` looks at `~/.local/Tachidesk` to write its setup. I let a comment not to forget what it's for

h7x4 commented:
> the program `tachidesk-server` looks at `~/.local/Tachidesk` to write its setup. I let a comment not to forget what it's for

Whoah, okay. Cool

> I put the changes in the correct commits

It's still a bit of a mess though. If you inspect https://github.com/NixOS/nixpkgs/commit/f41a3de4bdbeb5e2f0cd3fc3d4b60c815316c32d and [`1b8aa9e`](https://github.com/NixOS/nixpkgs/pull/257733/commits/1b8aa9ed133a71fb3714a36c0efac368dec3dc10), you'll see that they have a bunch of non-test and non-release-note related changes. There's also a small weirdness with introducing `nixosTests` and removing it again in the first two commits, but it's not as critical.

RatCornu commented:
> > the program `tachidesk-server` looks at `~/.local/Tachidesk` to write its setup. I let a comment not to forget what it's for
> 
> Whoah, okay. Cool
> 
> > I put the changes in the correct commits
> 
> It's still a bit of a mess though. If you inspect [f41a3de](https://github.com/NixOS/nixpkgs/commit/f41a3de4bdbeb5e2f0cd3fc3d4b60c815316c32d) and [`1b8aa9e`](https://github.com/NixOS/nixpkgs/pull/257733/commits/1b8aa9ed133a71fb3714a36c0efac368dec3dc10), you'll see that they have a bunch of non-test and non-release-note related changes. There's also a small weirdness with introducing `nixosTests` and removing it again in the first two commits, but it's not as critical.

Oh I'm sorry I'm not still really used to it  ^^', I try to fix this

RatCornu commented:
Okay I hope it's now correct

RatCornu commented:
Should I change something else or does everything is fine for you ?

h7x4 commented:
Hello again. I've been working on some changes that I thought would be painful to do through the review system. I think Sandros earlier comment about freeform was meant to hint at making the module in accordance with [`RFC-42`](https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md). This means including all settings that go into the config file into a `settings` attribute, and making that one freeform so that any non-declared settings also can be used with the module.

I've also added another test, just to verify that the program even runs when we do all the `basicAuthPassword` stuff.

Here's the patchset, hope you could take a look at it and maybe merge it into the commits: https://gist.github.com/h7x4/94185296658e58d3e5b93117aa9bc9a0. Feel free to ask questions if anything is unclear :grin:

RatCornu commented:
Okay I understand better the freeform type now. I read all the changes and everything looks fine. I didn't know `recursiveUpdate` but it looks very usefull. I will merge this in my commits as soon as possible!

RatCornu commented:
I'm sorry I think I did something I shouldn't have during a rebase...

h7x4 commented:
Please open a new PR :)
